### PR TITLE
SCP-2234: Chain index operations

### PR DESCRIPTION
**Pull request by j-mueller**
* Emulator implementations of chain index query and control effects
* Some property tests for the finger tree operations

Pre-submit checklist:
- Branch
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: plutus-chain-index/src/Plutus/ChainIndex/Tx.hs [hunk 1]
```diff
@@ -1,23 +1,91 @@
...
+            , _citxValidators = validatorHashes
+            }
+    Invalid tx@Tx{txCollateral, txValidRange, txData, txInputs, txMintScripts} ->
+        let (validatorHashes, otherDataHashes, redeemers) = validators txInputs in
+        ChainIndexTx
+            { _citxTxId = txId tx
+            , _citxInputs = txCollateral
+            , _citxOutputs = mempty
+            , _citxValidRange = txValidRange
+            , _citxData = txData <> otherDataHashes
```
Comment (luigy) [inline plutus-chain-index/src/Plutus/ChainIndex/Tx.hs:72]: not sure it if was intentional, but this used to be `mempty`?

Comment (j-mueller) [reply plutus-chain-index/src/Plutus/ChainIndex/Tx.hs:72]: Good spot, but this is intentional - the "disk state" part of the chain index is basically a content addressable store for pieces of Plutus data so, so we can include the datum hashes here even if the outputs don't make it into the ledger.

File: plutus-chain-index/src/Plutus/ChainIndex/Types.hs [hunk 2]
```diff
@@ -1,20 +1,66 @@
...
+-- | A page with the given size
+pageOf :: PageSize -> Set a -> Page a
+pageOf (PageSize ps) items =
+    let totalPages =
+            let (d, m) = Set.size items `divMod` ps'
+            in if m == 0 then d else d + 1
+        ps' = fromIntegral ps
+    in Page
+        { pageSize = PageSize ps
+        , pageNumber = 1
```
Comment (sjoerdvisscher) [inline plutus-chain-index/src/Plutus/ChainIndex/Types.hs:51]: When is `pageNumber` not equal to `1`?

File: plutus-chain-index/src/Plutus/ChainIndex/Emulator/Handlers.hs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+                logError $ Err reason
+                throwError reason
+            Right RollbackResult{newTip, rolledBackIndex} -> do
+                put $ oldState & set utxoIndex rolledBackIndex
+                logDebug $ RollbackSuccess newTip
+    CollectGarbage -> do
+        -- Rebuild the index using only transactions that still have at
+        -- least one output in the UTXO set
+        utxos <- gets (Set.toList . Set.map txOutRefId . UtxoState.unspentOutputs . UtxoState.utxoState . view utxoIndex)
+        newDiskState <- foldMap DiskState.fromTx . catMaybes <$> mapM getTxFromTxId utxos
```
Comment (sjoerdvisscher) [inline plutus-chain-index/src/Plutus/ChainIndex/Emulator/Handlers.hs:120]: Should it be an error if `getTxFromTxId` fails here?

File: plutus-chain-index/src/Plutus/ChainIndex/Emulator/Handlers.hs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    forall effs.
+    ( Member (State ChainIndexEmulatorState) effs
+    , Member (Error ChainIndexError) effs
+    ) => ChainIndexQueryEffect
+    ~> Eff effs
+handleQuery = \case
+    DatumFromHash h -> gets (view $ diskState . dataMap . at h)
+    ValidatorFromHash h -> gets (view $ diskState . validatorMap . at h)
+    MintingPolicyFromHash h -> gets (view $ diskState . mintingPolicyMap . at h)
+    TxOutFromRef TxOutRef{txOutRefId, txOutRefIdx} -> gets @ChainIndexEmulatorState (preview $ diskState . txMap . ix txOutRefId . citxOutputs . ix (fromIntegral txOutRefIdx))
```
Comment (sjoerdvisscher) [inline plutus-chain-index/src/Plutus/ChainIndex/Emulator/Handlers.hs:67]: Line is too long.

**Review by sjoerdvisscher — APPROVED** [review]

File: plutus-chain-index/src/Plutus/ChainIndex/Types.hs [hunk 5]
```diff
@@ -1,20 +1,66 @@
...
+        , pageItems = take ps' $ Set.toList items
+        }
+
+-- | The tip of the chain index
+data Tip =
+    -- TODO: Add genesis block tip
+    Tip
+        { tipSlot    :: Slot -- ^ Last slot
+        , tipBlockId :: BlockId -- ^ Last block ID
+        , tipBlockNo :: Int -- ^ Last block number (if available)
```
[the remark this answers is not in the document]
Comment (j-mueller) [reply plutus-chain-index/src/Plutus/ChainIndex/Types.hs:62]: ah yes, it's `Maybe Tip` now

File: plutus-chain-index/src/Plutus/ChainIndex/Types.hs (hunk 2 above)
Comment (j-mueller) [reply plutus-chain-index/src/Plutus/ChainIndex/Types.hs:51]: There isn't currently a way to request a page other than the first one (and I'm not 100% sure if this should even be possible), that's why it just returns the first page.

Maybe I should just change `Page` to `BoundedList` or something.

Comment (sjoerdvisscher) [reply plutus-chain-index/src/Plutus/ChainIndex/Types.hs:51]: The name `Page` does communicate the right idea. Just document this limitation with the `UtxoSetAtAddress` query.

File: plutus-chain-index/src/Plutus/ChainIndex/Emulator/Handlers.hs (hunk 3 above)
Comment (j-mueller) [reply plutus-chain-index/src/Plutus/ChainIndex/Emulator/Handlers.hs:138]: I changed `getTxFromTxId` to log a warning if the TX wasn't found

**Pull request merged**